src/Storages/System/StorageSystemDatabases.cpp

```diff
@@ -96,10 +98,43 @@ static ColumnPtr getFilteredDatabases(const Databases & databases, const Actions
     MutableColumnPtr engine_column = ColumnString::create();
     MutableColumnPtr uuid_column = ColumnUUID::create();
 
+    const auto access = context->getAccess();
+    const bool need_to_check_access_for_databases = !access->isGranted(AccessType::SHOW_DATABASES);
+
+    /// Optimization: For users with granular permissions, extract granted databases from access rights
+    /// to avoid calling isGranted() for each database (which is expensive for granular grants)
+    std::unordered_set<std::string> granted_databases;
+    bool has_wildcard_grant = false;
+
+    if (need_to_check_access_for_databases)
+    {
+        auto access_rights = access->getAccessRights();
+        auto elements = access_rights->getElements();
+
+        for (const auto & element : elements)
+        {
+            if (element.access_flags.contains(AccessType::SHOW_DATABASES))
```

Comment:
No, this approach won't work. The elements returned by `access_rights->getElements()` have more complicated structure: they can use partial revokes, for example
```
GRANT SHOW DATABASES ON *.*
REVOKE SHOW DATABASES ON db1.*
REVOKE SHOW DATABASES ON db2.*
```

Also wildcards can be used as a part of the database name, for example
```
GRANT SHOW DATABASES ON prefix1*.*
GRANT SHOW DATABASES ON prefix2*.*
```